cosign/toolchain.bzl

```diff
@@ -0,0 +1,78 @@
+"""This module implements the cosign-specific toolchain rule."""
+
+def _to_manifest_path(ctx, file):
+    if file.short_path.startswith("../"):
+        return "external/" + file.short_path[3:]
+    else:
+        return ctx.workspace_name + "/" + file.short_path
+
+CosignInfo = provider(
+    doc = "Information about how to invoke the cosign executable.",
+    fields = {
+        "cosign_path": "Path to the cosign executable for the target platform.",
+        "cosign_files": """Files required in runfiles to make the cosign executable available.
+
+May be empty if the cosign_path points to a locally installed tool binary.""",
+    },
+)
+
+def _cosign_toolchain_impl(ctx):
+    if ctx.attr.cosign and ctx.attr.cosign_path:
+        fail("Can only set one of cosign or cosign_path but both were set.")
+    if not ctx.attr.cosign and not ctx.attr.cosign_path:
+        fail("Must set one of cosign or cosign_path.")
+
+    cosign_files = []
+    cosign_path = ctx.attr.cosign_path
+
+    if ctx.attr.cosign:
+        cosign_files = ctx.attr.cosign.files.to_list()
+        cosign_path = _to_manifest_path(ctx, cosign_files[0])
+
+    # Make the $(COSIGN_BIN) variable available in places like genrules.
+    # See https://docs.bazel.build/versions/main/be/make-variables.html#custom_variables
+    template_variables = platform_common.TemplateVariableInfo({
+        "COSIGN_BIN": cosign_path,
+    })
+    default = DefaultInfo(
+        files = depset(cosign_files),
+        runfiles = ctx.runfiles(files = cosign_files),
+    )
+    cosign_info = CosignInfo(
+        cosign_path = cosign_path,
+        cosign_files = cosign_files,
```

Comment:
you're already passing cosign_files via the `default` attribute of `platform_common.ToolchainInfo` below so you don't need the `cosign_files`

should probably follow the same pattern as we did for YQ:

```
    # Make the $(YQ_BIN) variable available in places like genrules.
    # See https://docs.bazel.build/versions/main/be/make-variables.html#custom_variables
    template_variables = platform_common.TemplateVariableInfo({
        "YQ_BIN": binary.path,
    })
    default_info = DefaultInfo(
        files = depset([binary]),
        runfiles = ctx.runfiles(files = [binary]),
    )
    yq_info = YqInfo(
        bin = binary,
    )

    # Export all the providers inside our ToolchainInfo
    # so the resolved_toolchain rule can grab and re-export them.
    toolchain_info = platform_common.ToolchainInfo(
        yqinfo = yq_info,
        template_variables = template_variables,
        default = default_info,
    )
```

https://github.com/aspect-build/bazel-lib/blob/2689a48940007588e87f63acda704d72f2e2b6c2/lib/private/yq_toolchain.bzl#L111

Reply:
I didn't review the other toolchains in this rule set but perhaps we need to make similar changes in those as well

Reply:
ohhh.. if we're supporting a fixed `cosign_path` attribute in the toolchain however then we have no File object to pass 😞 

we don't have a yq_path option in the yq toolchain

Reply:
yq: https://github.com/aspect-build/bazel-lib/blob/2689a48940007588e87f63acda704d72f2e2b6c2/lib/private/yq_toolchain.bzl#L127

Reply:
How about passing the File object _or_ the path as a string either in the same field or in different fields. If its the same field the consumer will have to check if the value is a "string" and if it is treat it as an absolute path on disk.

Your `readonly COSIGN="$RUNFILES_ROOT/{{cosign_path}}"` will need adjusting as well if `{{cosign_path}}` is an absolute path.

Or for the mvp here, just do what yq does and keep it simple and don't allow passing an absolute path to the toolchain

Reply:
followed the approach of yq for the MVP. removed `chdir` behavior for now to keep it simple.